flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/FileChannelBoundedData.java

```diff
@@ -155,6 +159,10 @@ public void close() throws IOException {
 		@Override
 		public void recycle(MemorySegment memorySegment) {
 			buffers.addLast(memorySegment);
+
+			if (!isFinished) {
+				subpartitionView.notifyDataAvailable();
```

Comment:
Hmm, something is not right here.
1. there is a cyclic dependency between `FileBufferReader` and `BoundedBlockingSubpartitionReader`. Here we are calling `BoundedBlockingSubpartitionReader` which in turns call us back via `FileBufferReader#nextBuffer`
2. it's strange (when you say it out laud) that we are performing an io read in `recycle()` call stack
3. it's strange that `notifyDataAvailable` call is doing some io operation
4. it's also a bit strange that `recycle()` call is triggering `notifyDataAvailable()`, but this maybe unavoidable

for 2. and 3. names completely do not suggest what the methods are actually doing, and it is quite surprising that a light sounding method is doing some heavy work.

Can not we just notify the view, that we have memory to progress, and we would make io calls and maybe fill out read ahead buffer during next `BoundedBlockingSubpartitionReader#getNextBuffer` call? That would fix some of the problems that I mentioned: it would brake the cycle and there would be no io operations during recycle/data notification calls.

Reply:
In `recycle()` method, it only calls `notifyDataAvailable`, so we would not see io operation from this layer explicitly. So The key concern is that the implementation of `notifyDataAvailable` would be involved in io operation.

Actually `BoundedBlockingSubpartitionReader#getNextBuffer` and `notifyDataAvailable` are both called by netty thread, so it is no difference of io operation from caller aspect.

From the aspect of method naming semantic, it might seem a bit strange to make `notifyDataAvailable` do io operation. Actually I ever considered your suggested way during implementation, but there are some other problems if we do so:

- In `BoundedBlockingSubpartitionReader`, next buffer is also used for other places.  In the case of `nextBufferIsEvent` method, if we do not read ahead the next buffer during `notifyDataAvailable`, we do not know whether it is event or not.  Then in theory we still could not transport this event buffer if no credits.

- In the case of `BoundedBlockingSubpartitionReader#isAvailable()`, we still need to check whether the next buffer is null after `notifyDataAvailable` which keeps consistent with streaming mode and other types in batch mode. So we have to read ahead the next buffer before we actually notify view available to the netty stack.